plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -1,80 +1,244 @@
 -- Under-applied, not saturated
-saturatesScheme [] TypeSchemeArrow{}                     = Just False
-saturatesScheme [] TypeSchemeAll{}                       = Just False
+saturatesScheme AppContextEnd TypeSchemeArrow{}                 = Just False
+saturatesScheme AppContextEnd TypeSchemeAll{}                   = Just False
 -- These cases are only possible in case we have an ill-typed builtin application,
 -- so we can't give an answer.
-saturatesScheme (TypeArg _ : _) TypeSchemeArrow{}        = Nothing
-saturatesScheme (TermArg _ : _) TypeSchemeAll{}          = Nothing
+saturatesScheme TypeAppContext{} TypeSchemeArrow{}              = Nothing
+saturatesScheme TermAppContext{} TypeSchemeAll{}                = Nothing
 
--- | Is the given 'BuiltinApp' saturated?
--- Returns 'Nothing' if something is badly wrong and we can't tell.
+-- | Is the given application saturated?
+-- Returns 'Nothing' if we can't tell.
 isSaturated
     :: forall tyname name uni fun a
     . ToBuiltinMeaning uni fun
     => BuiltinSemanticsVariant fun
-    -> BuiltinApp tyname name uni fun a
+    -> fun
+    -> AppContext tyname name uni fun a
     -> Maybe Bool
-isSaturated semvar (BuiltinApp fun args) =
+isSaturated semvar fun args =
     case toBuiltinMeaning @uni @fun @(Term TyName Name uni fun ()) semvar fun of
         BuiltinMeaning sch _ _ -> saturatesScheme args sch
 
--- | View a 'Term' as a 'BuiltinApp' if possible.
-asBuiltinApp :: Term tyname name uni fun a -> Maybe (BuiltinApp tyname name uni fun a)
-asBuiltinApp = go []
-    where
-        go argsSoFar = \case
-            Apply _ t arg  -> go (TermArg arg:argsSoFar) t
-            TyInst _ t arg -> go (TypeArg arg:argsSoFar) t
-            Builtin _ b    -> Just $ BuiltinApp b argsSoFar
-            _              -> Nothing
+-- | Is this pure? Either yes, or maybe not.
+data Purity = MaybeImpure | Pure
 
-{- Note [Purity, strictness, and variables]
-Variables in PLC won't have effects: they can have something else substituted for them,
-but those will be fully evaluated already.
-However, in PIR we have non-strict variable bindings.
-References to non-strict variables *can* have effects - after all,
-they compile into an application.
+instance Pretty Purity where
+  pretty MaybeImpure = "impure?"
+  pretty Pure        = "pure"
 
-So we need to take this into account in our purity calculation.
-We require the user to tell us which variables are strict, this
-must be *conservative* (i.e. if you don't know, it's non-strict).
+-- | Is this term essentially work-free? Either yes, or maybe not.
+data WorkFreedom = MaybeWork | WorkFree
+
+instance Pretty WorkFreedom where
+  pretty MaybeWork = "maybe work?"
+  pretty WorkFree  = "work-free"
+
+-- | Either the "next" term to be evaluated, along with its 'Purity' and 'WorkFreedom',
+-- or we don't know what comes next.
+data EvalTerm tyname name uni fun a =
+  Unknown
+  | EvalTerm Purity WorkFreedom (Term tyname name uni fun a)
```

Comment:
I'd prefer to rule out the combination of `MaybeImpre` and `WorkFree` in the types, since an impure term cannot be work free.

Reply:
I'm actually not sure I agree with that! For example, I think it kind of makes sense to consider `error` to be impure but work-free.